### DISCUSSION TITLE
Best practice to restart onefile app

### DISCUSSION THREAD

**Discussion by jamesbcd** [pyinstaller/pyinstaller#6163] [Help]
Hi,

What's best practice for restarting a onefile-built app in a cross-platform way?

Background: My pyside2-based app uses a python plugin system (pluginlib). When plugins are removed, I'd like to be able to restart the app, to remove unused modules from the python interpreter.

Thanks for any tips you can pass on,

James.

**Comment by bwoodsend** [comment]
Honestly, I don't even know of a sensible way for Python to restart itself.

**Reply by aliencaocao** [reply]
You can use

`os.system('python your this script.py')
exit()
`
It works for me i've been using it

**Reply by bwoodsend** [reply]
That uses your original Python installation which defeats the purpose of PyInstaller. If you run your program on a machine which doesn't have Python installed then it will break.

**Comment by debilin** [comment]
Is there any solution for this problem?

**Comment by dennisvang** [comment]
Maybe something like the following, with the help of [`sys.executable`][1]?
(looking hopefully @bwoodsend for confirmation)

```python
if hasattr(sys, 'frozen'):
    subprocess.Popen(sys.executable)
    sys.exit(0)
```
On windows this works without `shell=True`, but I'm not sure about other OSes. Also see [subprocess docs][2]

[1]: https://pyinstaller.org/en/stable/runtime-information.html?highlight=sys.executable#using-sys-executable-and-sys-argv-0
[2]: https://docs.python.org/3/library/subprocess.html#popen-constructor

**Reply by bwoodsend** [reply]
It seems to work on UNIX and Windows although I'm surprised that it does. My understanding was that any subprocess is terminated whenever the parent process quits unless you disown the subprocess (which is remarkably fiddly to do) – that's an OS constraint and not even Python specific – so I was expecting either the old process to not quit or the new process to shut down as soon as the old one does.

**Reply by aliencaocao** [reply]
Its not killed in this case as the main process dont share the same resource as the new process. Its unlike using python multiprocessing, since in this case your python interpreter is different.

**Reply by dennisvang** [reply]
On Windows, `subprocess` uses `CreateProcess` ([source][1]), which, [according to MS][2]:

>... creates a new process that runs independently of the creating process.

(although "independently" is open to interpretation)

On posix, [`subprocess`][5] uses either [`fork()`][6] ([source][4]), `vfork()`, or [`posix_spawn()`][7]. 

If I understand correctly, the discussions [here][8] and [here][9] suggest that forked child processes typically stay alive after their parent process dies, unless you make an effort to kill them.

[1]: https://github.com/python/cpython/blob/351c103134e43c2ee43deb10cdc9afb37b916a4e/Lib/subprocess.py#L1543
[2]: https://learn.microsoft.com/en-us/windows/win32/procthread/creating-processes
[3]: https://github.com/python/cpython/blob/351c103134e43c2ee43deb10cdc9afb37b916a4e/Lib/subprocess.py#L1755
[4]: https://github.com/python/cpython/blob/351c103134e43c2ee43deb10cdc9afb37b916a4e/Modules/_posixsubprocess.c#L923
[5]: https://github.com/python/cpython/blob/main/Lib/subprocess.py
[6]: https://man7.org/linux/man-pages/man2/fork.2.html
[7]: https://man7.org/linux/man-pages/man3/posix_spawn.3.html
[8]: https://unix.stackexchange.com/q/158727
[9]: https://unix.stackexchange.com/q/149319

**Reply by bwoodsend** [reply]
Note that in order to clear up some of the mess surrounding multiprocessing, we had to add an extra hoop to jump through. See [PYINSTALLER_RESET_ENVIRONMENT](https://pyinstaller.org/en/v6.10.0/common-issues-and-pitfalls.html#independent-subprocess) for PyInstaller >= v6.10.
